cli/upgrade.rs

```diff
@@ -0,0 +1,227 @@
+  }
+  Ok(())
+}
+
+fn download_package(
+  url: &Url,
+  client: Client,
+) -> Pin<Box<dyn Future<Output = Result<Vec<u8>, ErrBox>>>> {
+  println!("downloading {}", url);
+  let url = url.clone();
+  let fut = async move {
+    match fetch_once(client.clone(), &url, None).await? {
+      FetchOnceResult::Code(source, _) => Ok(source),
+      FetchOnceResult::NotModified => unreachable!(),
+      FetchOnceResult::Redirect(_url, _) => {
+        download_package(&_url, client).await
+      }
+    }
+  };
+  fut.boxed_local()
+}
+
+fn compose_url_to_exec(version: &Version) -> Result<Url, ErrBox> {
+  let s = format!(
+    "https://github.com/denoland/deno/releases/download/v{}/{}",
+    version, ARCHIVE_NAME
+  );
+  Ok(Url::parse(&s)?)
+}
+
+fn find_version(text: &str) -> Result<String, ErrBox> {
+  let re = Regex::new(r#"v([^\?]+)?""#)?;
+  if let Some(_mat) = re.find(text) {
+    let mat = _mat.as_str();
+    return Ok(mat[1..mat.len() - 1].to_string());
+  }
+  Err(ErrorMsg("Cannot read latest tag version".to_string()).to_err_box())
+}
+
+fn unpack(archive_data: Vec<u8>) -> Result<PathBuf, ErrBox> {
+  // We use into_path so that the tempdir is not automatically deleted. This is
+  // useful for debugging upgrade, but also so this function can return a path
+  // to the newly uncompressed file without fear of the tempdir being deleted.
+  let temp_dir = TempDir::new()?.into_path();
+  let exe_ext = if cfg!(windows) { "exe" } else { "" };
+  let exe_path = temp_dir.join("deno").with_extension(exe_ext);
+  assert!(!exe_path.exists());
+
+  let archive_ext = Path::new(ARCHIVE_NAME)
+    .extension()
+    .and_then(|ext| ext.to_str())
+    .unwrap();
+  let unpack_status = match archive_ext {
+    "gz" => {
+      let exe_file = fs::File::create(&exe_path)?;
+      let mut cmd = Command::new("gunzip")
+        .arg("-c")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::from(exe_file))
+        .spawn()?;
+      cmd.stdin.as_mut().unwrap().write_all(&archive_data)?;
+      cmd.wait()?
+    }
+    "zip" => {
+      if cfg!(windows) {
+        let archive_path = temp_dir.join("deno.zip");
+        fs::write(&archive_path, &archive_data)?;
+        Command::new("powershell.exe")
+          .arg("-Command")
+          .arg("Expand-Archive")
+          .arg("-Path")
+          .arg(&archive_path)
+          .arg("-DestinationPath")
+          .arg(&temp_dir)
+          .spawn()?
+          .wait()?
+      } else {
+        let archive_path = temp_dir.join("deno.zip");
+        fs::write(&archive_path, &archive_data)?;
+        Command::new("unzip")
+          .current_dir(&temp_dir)
+          .arg(archive_path)
+          .spawn()?
+          .wait()?
+      }
+    }
+    ext => panic!("Unsupported archive type: '{}'", ext),
+  };
+  assert!(unpack_status.success());
+  assert!(exe_path.exists());
+  Ok(exe_path)
+}
+
+fn replace_exe(new: &Path, old: &Path) -> Result<(), ErrBox> {
+  fs::remove_file(old)?;
```

Comment:
* On Unix systems, removing the old file is unnecessary. `rename()` will replace an existing file with the same name.
* On Windows, you cannot overwrite or delete a running executable. The trick is to rename the executable first.

I think this will work:
```
if cfg!(windows) {
  let delete_me = old.with_extension("delete_me");
  fs::rename(old, &delete_me)?;
}
fs::rename(new, old)?;
```